### PR TITLE
DO NOT MERGE: Added pyheron API + basic component for python topology support

### PR DESCRIPTION
**Pull request by taishi8117**
This is a working branch and these files are NOT yet ready to be merged; this PR is opened for reviewing purposes. I will need to add unit tests and codes for remaining functionalities, as well as integration tests once other python topology components are ready, before this PR is merged.

These files are Heron Instance's base components (spout/bolt/topology) in Python. Modules included in this PR serve similar functionalities as ``BoltInstance``, ``SpoutInstance``, API interfaces for Spout and Bolt, ``TopologyBuilder`` and ``HeronSubmitter`` etc. in Java implementation. 

Once Python topology is officially supported, these files will be packaged into a library (maybe called ``pyheron``) for topology development; they are also made compatible with [StreamParse API](http://streamparse.readthedocs.io/en/v3.0.0/api.html).

### CODE REVIEW

**Comment by amontalenti** [comment]
Hey! One of the streamparse original creators here. This is looking really interesting so far! I added a comment on #574 about recent happenings in pystorm & streamparse. But I agree with the approach you've taken of building your own pyheron interop layer and then just matching the duck-typed API that streamparse's Topology DSL and that pystorm components implement.

**Comment by taishi8117** [comment]
@amontalenti Thanks for your comment! As you suggested, we are making the Heron Python topology as compatible with streamparse API (v3.0) as possible. It would also be great if you could take a look at our codes and give us some suggestions.

File: heron/instance/src/python/basics/component.py [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+    self.output_helper.add_control_tuple(control_tuple, tuple_size_in_bytes, is_ack)
+
+  def get_total_data_emitted_in_bytes(self):
+    return self.output_helper.total_data_emitted_in_bytes
+
+  ##################################################################
+  # The followings are to be implemented by Spout/Bolt independently
+  ##################################################################
+
+  def start(self):
```
Comment (objmagic) [inline heron/instance/src/python/basics/component.py]: Please use ``@abstractmethod`` decorator.

**Comment by taishi8117** [comment]
API (#1236) is separated from instance, will close this PR.

**Pull request closed without merging**